### DISCUSSION TITLE
Feature set and use cases

### DISCUSSION THREAD

**Discussion by zigmhount** [SanderVocke/shoopdaloop#301] [Ideas]
Hello again :)

Continuing the conversation we started in #235 :

> At the moment I am finding that instead of making music, which was the point, I am mostly working on this project instead. Ah well, at some point the balance will shift, and meanwhile working on this is also fun. 

Haha, same here! But I eventually came to accept that I may be more an engineer than a musician, always finding ideas to improve the tools even though I'm not actually using the tools much :sweat_smile: 

> Mostly I am wondering if there are any major gaps in terms of features (including to-do features already listed in the issue list), that would make this a dealbreaker. Also curious for the ideas you had for your own project - who knows, that might be inspiring.

That said, if you haven't, I would advise you to try and get to learn some of the commercial alternatives - Bitwig is the only one I know that can do what we're looking for on Linux - since there are great engineers there that have been paid for many years to figure out the best workflow and interface. I used it for a few months after having worked on my own [audio looping project](https://github.com/zigmhount/pdlooper) built in Pure Data, and a few things (including opinionated restrictions) were quite enlightening.

My latest project has been to use [Seq192](https://github.com/jean-emmanuel/seq192) as a MIDI looper and [Non-Mixer XT](https://github.com/Stazed/non-mixer-xt) as plugin host and mixer, hosted together in a [RaySession](https://github.com/Houston4444/RaySession) issue where I would include a few other Python scripts or Pure Data patches to use some custom MIDI instruments. I'm controlling Seq192 with an APC mini (hence currently writing a Python program to convert MIDI to OSC and OSC to LED colors), and a Launch Control XL to manipulate the plugins and effects in Non-Mixer XT (I used Ardour before, where the MIDI learn and controls binding is much smoother and more reliable).

Off the top of my head, these would be the most critical features for my own use:
- Duplicate loop #241 (since in many cases I would reuse the drums track for several scenes)
- High CPU usage #125 
- Binding MIDI controllers to the hosted plugins' controls. Not sure how doable this is through Carla.
- Triggering rec/stop with another MIDI controller (e.g. a pedal, while my hands are taken on a keyboard). Probably already possible, I'll try asap.
- Grouping tracks (e.g. for scenes or drumsets) which may already doable but I haven't figured out yet how to use "Trigger-only" tracks
- Not a deal-breaker at first, but nice and fun to play with: Auto-follow, e.g. automatically trigger another loop in the track after X loops of the current one. Bitwig and Live have this, where you can select e.g. "Next" (=the loop below the current one), "Previous", "None" (=stop after this loop), "Random" (=any loop in the same track), and even selecting a set of loops that can follow the selected one.
- Nice to have: exporting all MIDI files for all loops at once. One of my latest goals is to export MIDI, reimport it into MuseScore (or Lilypond) and produce a score of my session, so that I can easily "live-reproduce" it instead of playing pre-recorded stuff of improvising from scratch every time.

**Comment by SanderVocke** [comment]
> That said, if you haven't, I would advise you to try and get to learn some of the commercial alternatives - Bitwig is the only one I know that can do what we're looking for on Linux - since there are great engineers there that have been paid for many years to figure out the best workflow and interface. I used it for a few months after having worked on my own [audio looping project](https://github.com/zigmhount/pdlooper) built in Pure Data, and a few things (including opinionated restrictions) were quite enlightening.

That's good to know - until now I assumed that only Ableton would fit the bill. Having a Linux alternative is nice. Why did you stop using that BTW?

> My latest project has been to use [Seq192](https://github.com/jean-emmanuel/seq192) as a MIDI looper and [Non-Mixer XT](https://github.com/Stazed/non-mixer-xt) as plugin host and mixer, hosted together in a [RaySession](https://github.com/Houston4444/RaySession) issue where I would include a few other Python scripts or Pure Data patches to use some custom MIDI instruments. I'm controlling Seq192 with an APC mini (hence currently writing a Python program to convert MIDI to OSC and OSC to LED colors), and a Launch Control XL to manipulate the plugins and effects in Non-Mixer XT (I used Ardour before, where the MIDI learn and controls binding is much smoother and more reliable).

I really like the idea of multi-app session management, and originally I intended for ShoopDaLoop to be used this way (and it still can, with the "external" dry-wet tracks). Unfortunately, for me personally I had lots of issues with RaySession and Ajordejo that made it unworkeable. Probably my own fault, as I hear much good about these tools.

One drawback is that by doing a dry/wet setup, you have to make a cycle in your Jack graph, which effectively doubles the total latency.

>     * Duplicate loop [Duplicate loop #241](https://github.com/SanderVocke/shoopdaloop/issues/241) (since in many cases I would reuse the drums track for several scenes)

This one is high on my list!

FYI though: have a play with the composite loops, because it sounds like what you want is already possible. Shoop doesn't have "scenes" but only "composite loops", which can be used as a scene or a repeating sequence or both. You can re-use your single drum loop in multiple composite loops.
The feature is pretty recent and I still need to improve the documentation. But essentially, just right-click any loop, click "create composite", open the details pane in the bottom and drag other loops into the timeline.

>     * Binding MIDI controllers to the hosted plugins' controls. Not sure how doable this is through Carla.

I am also not sure yet how to tackle this. But agree it is important.

>     * Triggering rec/stop with another MIDI controller (e.g. a pedal, while my hands are taken on a keyboard). Probably already possible, I'll try asap.

Indeed, this is possible.
If you assign the "default loop action" to your MIDI controller, that would toggle between rec, play, stop on a single trigger. Or you can map specific commands.

>     * Grouping tracks (e.g. for scenes or drumsets) which may already doable but I haven't figured out yet how to use "Trigger-only" tracks

Indeed, a lot is already possible with the composite loops. Be aware: you don't need a trigger-only track for it. Any loop can be a composite loop. For example, if I want to sequence loops of the same track, I usually put the composite loop in the same track. When I want to group stuff over multiple tracks, I like to add those composites to a trigger-only track to keep the overview.

>     * Not a deal-breaker at first, but nice and fun to play with: Auto-follow, e.g. automatically trigger another loop in the track after X loops of the current one. Bitwig and Live have this, where you can select e.g. "Next" (=the loop below the current one), "Previous", "None" (=stop after this loop), "Random" (=any loop in the same track), and even selecting a set of loops that can follow the selected one.

That sounds cool. In principle you can already do it with the composite loops, but you'd have to manually add each new recording to the list, as opposed to it just following the order of the track. I'll have a think about it.

The "random" part is also nice! I feel like instead of supporting these kind of things directly, what I'd really like is if the LUA scripting interface becomes powerful enough so that users can program their own fun behaviors. But it would mainly require having a way to couple these scripts to buttons or loops.

For example, I went the same route with the keyboard controls. I wanted to offer comprehensive controls but also give the opportunity for users to create their own keyboard mappings and advanced keyboard behaviors. So the built-in keyboard support is a LUA script shipped with the app, which the user can copy-paste and modify to their taste if they prefer.

It's just that the LUA script interface is relatively limited ATM. I will expand it as new scripting ideas come up. It won't be a stable interface until a first release comes.

>     * Nice to have: exporting all MIDI files for all loops at once. One of my latest goals is to export MIDI, reimport it into MuseScore (or Lilypond) and produce a score of my session, so that I can easily "live-reproduce" it instead of playing pre-recorded stuff of improvising from scratch every time.

Thanks for your insights!

**Comment by SanderVocke** [comment]
As for binding plugin settings to MIDI controls: there is currently no specific feature for it. From what I know, Carla can already bind controls to parameters though. The downside is that the information can only get in through the same MIDI port that the recording goes into.

In other words, if you would want to control a plugin this way, you'd have to connect that MIDI controller to the recording input of the track. And any control changes to that parameter would be recorded along with the content (and played back again).

Edit: #324

**Comment by zigmhount** [comment]
Sorry for the silence, it's been a busy couple of weeks. Here are finally a few follow-up answers:

> Why did you stop using that [Bitwig] BTW?

I was more excited by the idea of trying to get a similar behaviour with open-source only - and also, the free plan (not sure if it still exists) lets you use only 8 tracks, only 2 or 3 external plugins, and no major version upgrades. The price then gets quite steep.

> Unfortunately, for me personally I had lots of issues with RaySession and Ajordejo that made it unworkeable. Probably my own fault, as I hear much good about these tools.

Tbh I'm struggling quite a lot too, especially with the connections and the MIDI controls mapping, and Non-Mixer(-XT)'s midi-mapper (midi-osc interface). Sometimes I feel like giving up (which I also dis in the past haha) or writing a simple script to just launch and connect everything the way I want it - which I may well do soon.

> have a play with the composite loops

Aha, interesting, I'll try that!

> Be aware: you don't need a trigger-only track for it.

Good to know, but what are trigger-only tracks for then? To start one-shot effects/samples?

> what I'd really like is if the LUA scripting interface becomes powerful enough so that users can program their own fun behaviors. But it would mainly require having a way to couple these scripts to buttons or loops.

Sounds very cool with a button indeed. I'll have to get into LUA scripting.

> From what I know, Carla can already bind controls to parameters though.

Yeah, it is possible, but I find it very cumbersome and unintuitive to (IIRC) select the parameter in a long list (as opposed to clicking on the control in the UI) and then select the control in a long list (as opposed to moving the physical knob).
I'm at a stage where I don't know in advance which parameters are useful or will be fun to change during the performance, so I really like the workflow "activate midi learn - touch a parameter - move the controller's knob - done" and it's important that it stays reliable (hence my troubles with Non-Mixer) (I even had implemented midi learn in one of my Pure Data projects).

Anyway, thanks for you continued work on this, I hope I'll find more time to play with shoop in 10 days or so.

**Comment by SanderVocke** [comment]
> > Be aware: you don't need a trigger-only track for it.
> 
> Good to know, but what are trigger-only tracks for then? To start one-shot effects/samples?
> 

They don't add any particular functionality. I just added them because I think as a session becomes a bit complex, you may want to have a track with only composite loops in it, for "organizing" your session. For example, you can make a trigger-only track, and make each loop in it connect to all others in the same row - that way you'd have created something like Ableton's "scenes".

> > what I'd really like is if the LUA scripting interface becomes powerful enough so that users can program their own fun behaviors. But it would mainly require having a way to couple these scripts to buttons or loops.
> 
> Sounds very cool with a button indeed. I'll have to get into LUA scripting.

Cool! The groundwork is there, but the LUA API is quite limited. I've just been adding functions to that as I needed them, mostly for the default keyboard script. But I think it can become quite powerful.

> > From what I know, Carla can already bind controls to parameters though.
> 
> Yeah, it is possible, but I find it very cumbersome and unintuitive to (IIRC) select the parameter in a long list (as opposed to clicking on the control in the UI) and then select the control in a long list (as opposed to moving the physical knob). I'm at a stage where I don't know in advance which parameters are useful or will be fun to change during the performance, so I really like the workflow "activate midi learn - touch a parameter - move the controller's knob - done" and it's important that it stays reliable (hence my troubles with Non-Mixer) (I even had implemented midi learn in one of my Pure Data projects).

That's a shame, I personally haven't played with it yet. At the moment I've tied everything to Carla pretty hard for plugin hosting, so if Carla is not intuitive in that regard, I won't be able to work around it. The only option would be to use external apps (via JACK/NSM) instead of hosted plugins via Carla, that way you can do it however you prefer.

> Anyway, thanks for you continued work on this, I hope I'll find more time to play with shoop in 10 days or so.

Cheers!

Also a FYI: the next build I will upload will have quite some changes to the loop controls. I reduced it back to "play" "play dry", "record", "grab" (new), "stop".
There are now global controls at the top for "solo" (if active, e.g. pressing play will stop other loops in the track, but also works for record now) and "play after record" (if active, recording a loop for a fixed amount of time will play it afterward) and "n cycles" (the amount of cycles to record).
It is mostly a restructuring of the already supported behaviors. These global controls are also mapped to Ctrl, Alt and Shift.

The "grab" is new: I am working on always-on recording. So you don't have to press anything to record - just play (even with the ear symbol off - no monitoring). When you play something nice, hit "grab" (G). The grab respects the N cycles control and the "play after recording" control (will play immediately in that case).

I will add documentation soon, but just a heads-up in the meantime.

**Reply by SanderVocke** [reply]
New build is up.

**Comment by SanderVocke** [comment]
FYI: I updated the [documentation](https://docs.shoopdaloop.com). The "Controlling Loops" section now explains the new global controls in more detail, and also goes into the "grabbing" (always-on recording) feature.

I feel like it has reached the point now that these basic controls will stay relatively stable, so it really is time to make a demo video soon!
